app/queries/servers/channel.ts

```diff
@@ -431,3 +432,99 @@ export function queryMyChannelsByUnread(database: Database, isUnread: boolean, s
+    const value = Q.sanitizeLikeString(term.startsWith('@') ? term.substring(1) : term);
+    let username = `u.username like '${value}%'`;
+    let displayname = `c.display_name like '${value}%'`;
+    if (!matchStart) {
+        username = `u.username like '%${value}%' and u.username not like '${value}%'`;
+        displayname = `(c.display_name like '%${value}%' and c.display_name not like '${value}%')`;
+    }
+    const currentUserId = observeCurrentUserId(database);
+    return currentUserId.pipe(
+        switchMap((uId) => {
+            return database.get<MyChannelModel>(CHANNEL).query(
+                Q.unsafeSqlQuery(`select distinct my.* from ${MY_CHANNEL} my
+                inner join ${CHANNEL} c on c.id=my.id and c.team_id='' and c.delete_at=0 ${onlyDMs}
+                inner join ${CHANNEL_MEMBERSHIP} cm on cm.channel_id=my.id
+                inner join ${USER} u on u.id=cm.user_id and (cm.user_id != '${uId}' and ${username})
+                or ${displayname}
+                order by my.last_viewed_at desc
+                limit ${take}`),
+            ).observe();
+        }),
+    );
+};
+
+export const observeNotDirectChannelsByTerm = (database: Database, term: string, take = 20, matchStart = false) => {
+    const teammateNameSetting = observeTeammateNameDisplay(database);
+
+    const value = Q.sanitizeLikeString(term.startsWith('@') ? term.substring(1) : term);
+    let username = `u.username like '${value}%'`;
+    let nickname = `u.nickname like '${value}%'`;
+    let displayname = `(u.first_name || ' ' || u.last_name) like '${value}%'`;
+    if (!matchStart) {
+        username = `(u.username like '%${value}%' and u.username not like '${value}%')`;
+        nickname = `(u.nickname like '%${value}%' and u.nickname not like '${value}%')`;
+        displayname = `((u.first_name || ' ' || u.last_name) like '%${value}%' and (u.first_name || ' ' || u.last_name) not like '${value}%')`;
+    }
+
+    return teammateNameSetting.pipe(
+        switchMap((setting) => {
+            let sortBy = '';
+            switch (setting) {
+                case General.TEAMMATE_NAME_DISPLAY.SHOW_NICKNAME_FULLNAME:
+                    sortBy = "order by case u.nickname when '' then 1 else 0 end, case (u.first_name || ' ' || u.last_name) when '' then 1 else 0 end, u.username";
+                    break;
+                case General.TEAMMATE_NAME_DISPLAY.SHOW_FULLNAME:
+                    sortBy = "order by case (u.first_name || ' ' || u.last_name) when '' then 1 else 0 end, u.username";
+                    break;
+                default:
+                    sortBy = 'order u.username';
+                    break;
+            }
+
+            return database.get<UserModel>(USER).query(
+                Q.unsafeSqlQuery(`select distinct u.* from User u
+                left join ChannelMembership cm ON cm.user_id=u.id
+                where cm.user_id IS NULL and (${displayname} or ${username} or ${nickname})
```

Comment:
I may be missing something here. If cm.user_id = u.id, why the where we check cm.user_id IS NULL?

Reply:
Right, let's see if this helps:

At the DB level we are selecting all rows from the User table and for each row we attempt to find a row in ChannelMembership that matches on the `id` column. If there is no such row, we just leave the `ChannelMembership` portion of the result as empty for that row. Then we constrain the selection by picking only those rows in the result where the matching row does not exists and finally return only the fields from the user table.